modules/trainer/GenericTrainer.py

```diff
@@ -319,6 +322,63 @@ def __sample_during_training(
 
         torch_gc()
 
+    def __validate(self, train_progress):
+        self.validation_data_loader.get_data_set().start_next_epoch()
+        current_epoch_length_validation = self.validation_data_loader.get_data_set().approximate_length()
+        if current_epoch_length_validation > 0 and self.__needs_validate(train_progress):
+            torch_gc()
+
+            step_tqdm_validation = tqdm(
+                self.validation_data_loader.get_data_loader(),
+                desc="validation_step",
+                total=current_epoch_length_validation,
+                initial=0)
+
+            accumulated_loss_per_concept = {}
+            concept_counts = {}
+            mapping_seed_to_label = {}
+            mapping_label_to_seed = {}
+
+            for epoch_step_validation, validation_batch in enumerate(step_tqdm_validation):
+
+                if self.__needs_gc(train_progress):
+                    torch_gc()
+
+                with torch.no_grad():
+                    model_output_data = self.model_setup.predict(self.model, validation_batch, self.config,
+                                                                 train_progress)
+                    loss_validation = self.model_setup.calculate_loss(self.model, validation_batch, model_output_data,
+                                                                      self.config, mean=False)
```

Comment:
I do not understand why you would want the mean to be false. Validation / test set loss is generally supposed to be performed using the exact same mechanics as training set loss. That includes taking the mean over your batch sizes. If you don't do that, you can't measurably compare the difference between training and test set losses.

(Also removing it will make the code simpler as your above tweaks to allow mean to be false at all wouldn't need to happen.)

Reply:
We want to calculate a separate validation loss for each concept, and each batch can contain samples of multiple concepts. Without the mean, we can separate the individual losses and average at the end.

But there is actually another problem. With batching it's possible that some validation samples are dropped, making the loss hard to compare between epochs. Maybe we need to set the batch size to 1

Reply:
reverted this moment with mean flag and set batch size = 1